contracts/RewardController.sol

```diff
@@ -0,0 +1,252 @@
+            vaultInfo[_vault].lastRewardBlock = block.number > startBlock ? block.number : startBlock;
+            return;
+        }
+        if (block.number <= lastRewardBlock) {
+            return;
+        }
+
+        vault.lastRewardBlock = block.number;
+
+        uint256 totalShares = IERC20Upgradeable(_vault).totalSupply();
+
+        if (totalShares != 0) {
+            uint256 reward = getVaultReward(_vault, lastRewardBlock);
+            vault.rewardPerShare += (reward * 1e12 / totalShares);
+        }
+
+        if (globalVaultsUpdates.length != 0) {
+            vaultInfo[_vault].lastProcessedTotalAllocPoint = globalVaultsUpdates.length - 1;
+        }
+    }
+
+    /**
+     * @notice Called by owner to set reward multipliers
+     * @param _rewardPerEpoch reward multipliers
+    */
+    function setRewardPerEpoch(uint256[24] memory _rewardPerEpoch) external onlyOwner {
+        rewardPerEpoch = _rewardPerEpoch;
+        emit SetRewardPerEpoch(_rewardPerEpoch);
+    }
+
+    function _updateVaultBalance(
+        address _vault,
+        address _user,
+        uint256 _sharesChange,
+        bool _isDeposit
+    ) internal {
+        updateVault(_vault);
+
+        uint256 userShares = IERC20Upgradeable(_vault).balanceOf(_user);
+        uint256 rewardPerShare = vaultInfo[_vault].rewardPerShare;
+        unclaimedReward[_vault][_user] += userShares * rewardPerShare / 1e12 - rewardDebt[_vault][_user];
+
+        if (_sharesChange != 0) {
+            if (_isDeposit) {
+                userShares += _sharesChange;
+            } else {
+                userShares -= _sharesChange;
+            }
+        }
+        rewardDebt[_vault][_user] = userShares * rewardPerShare / 1e12;
+    }
+
+    function updateVaultBalance(address _user, uint256 _sharesChange, bool _isDeposit) external {
+        _updateVaultBalance(msg.sender, _user, _sharesChange, _isDeposit);
+    }
+
+    /**
+     * @notice Transfer to the specified user their pending share of rewards.
+     * @param _vault The vault address
+     * @param _user The user address to claim for
+     */
+    function claimReward(address _vault, address _user) external {
+        _updateVaultBalance(_vault, _user, 0, true);
+
+        uint256 userUnclaimedReward = unclaimedReward[_vault][_user];
+        if (userUnclaimedReward > 0) {
+            unclaimedReward[_vault][_user] = 0;
+            safeTransferReward(_user, userUnclaimedReward, _vault);
+        }
+
+        emit ClaimReward(_vault, _user, userUnclaimedReward);
+    }
+
+     /**
+    * @notice Calculate rewards for a vault by iterating over the history of totalAllocPoints updates,
+    * and sum up all rewards periods from vault.lastRewardBlock until current block number.
+    * @param _vault The vault address
+    * @param _fromBlock The block from which to start calculation
+    * @return reward
+    */
+    function getVaultReward(address _vault, uint256 _fromBlock) public view returns(uint256 reward) {
```

Comment:
seems that this would revert if _fromBlock < startBlock or if fromBlock is from before the lastProcessedTotalAllocpoint

if so, this is apparently not a problem for internal calls, but  is still unexpected behavior that can be handled easily